unity-client/Assets/Scripts/MainScripts/DCL/Controllers/BuilderMode/Catalog/SceneObjectCatalogController.cs

```diff
@@ -0,0 +1,389 @@
+                        {
+                            foundCategory = true;
+                            if (!groupedSceneObjects[sceneObject.category].Contains(sceneObject))
+                                groupedSceneObjects[sceneObject.category].Add(sceneObject);
+                        }
+                    }
+                    if (!foundCategory)
+                    {
+                        Dictionary<string, List<SceneObject>> groupedSceneObjects = new Dictionary<string, List<SceneObject>>();
+                        groupedSceneObjects.Add(sceneObject.category, new List<SceneObject>() { sceneObject });
+                        filterObjects.Add(groupedSceneObjects);
+                    }
+                }
+            }
+        }
+    }
+
+    #endregion
+
+    #region DragAndDrop
+
+    int lastIndexDroped = -1;
+    GameObject draggedObject;
+    public void SetIndexToDrop(int index)
+    {
+        lastIndexDroped = index;
+    }
+
+    void OnDrag(PointerEventData data)
+    {
+        draggedObject.transform.position = data.position;
+    }
+
+    void SceneObjectStartDragged(SceneObject sceneObjectClicked, CatalogItemAdapter adapter, BaseEventData data)
+    {
+        PointerEventData eventData = data as PointerEventData;
+        if(draggedObject== null)
+            draggedObject = Instantiate(adapter.gameObject, generalCanvas.transform);
+        RectTransform adapterRT = adapter.GetComponent<RectTransform>();
+        RectTransform newAdapterRT = draggedObject.GetComponent<RectTransform>();
+        CatalogItemAdapter newAdapter = draggedObject.GetComponent<CatalogItemAdapter>();
+        newAdapter.canvasGroup.blocksRaycasts = false;
+        newAdapter.canvasGroup.alpha = 0.6f;
+        newAdapter.SetContent(adapter.GetContent());
+        newAdapterRT.sizeDelta = adapterRT.sizeDelta*0.75f;
+        OnStopInput?.Invoke();
+    }
+
+    void OnEndDrag(PointerEventData data)
+    {     
+        Destroy(draggedObject,0.1f);
+        OnResumeInput?.Invoke();
+    }
+
+    public void SceneObjectDropped(BaseEventData data)
+    {
+     
+        CatalogItemAdapter adapter = draggedObject.GetComponent<CatalogItemAdapter>();
+        SceneObject sceneObject = adapter.GetContent();
+        Texture texture = null;
+        if (adapter.thumbnailImg.enabled)
+        {
+            texture = adapter.thumbnailImg.texture;
+            SetQuickBarShortcut(sceneObject, lastIndexDroped, texture);
+        }
+        Destroy(draggedObject);
+    }
+
+    #endregion
+
+    void SetQuickBarShortcut(SceneObject sceneObject, int index,Texture texture)
+    {     
+        quickBarShortcutsSceneObjects[index] = sceneObject;
+        if (index < shortcutsImgs.Length)
+        {
+            shortcutsImgs[index].texture = texture;
+            shortcutsImgs[index].enabled = true;
+        }
+    }
+
+    #region Favorites
+
+    public void ShowFavorites()
```

Comment:
Im also a bit concerned about the performance of this method (these 4-depth foreach nest is scary) . rebuilding the favorites list every time seems like a waste. Cannot we build it once and then add/remove items?

Also I think the approach could be simplified. It would be easier to just carry a list of IDs as "favorites" and when adding/removing update the shortcut items (just resolving by ID) and that's it. The heart on each item is enabled/disabled checking against a simple:
` public bool IsFavorite(string itemId) => return favorites.Contains(itemId);`

Reply:
The performance shouldn't be a problem since this function is called only once each time that the user wants to see all of its favorites, 
The problem that I see is that if the scene assets pack changes, the favorites will show items that are not currently available. If you add ( in the future) an asset pack, added a favorite, and then remove the asset pack, you will have the favorite but you will have a desync between the current asset packs and the favorites objects. I don't know if that is the expected behavior

However, as state before, this whole class needs refactoring once the purpose of the whole system is clear.

Reply:
I dont see that as a blocker for the suggested approach. A change in the asset pack can just trigger a cleanup of the favorites that are part of that pack.

In any case we should, indeed, move this to another PR

Reply:
Added to the document for future revision